modules/sources/contracts/fetch-pool-data.ts

```diff
@@ -1,95 +1,97 @@
+    blockNumber?: bigint,
+): Promise<{ [address: string]: OnchainPoolData }> {
+    const contracts = pools
         .map((pool) => [
             {
                 address: vault as `0x${string}`,
                 abi: vaultV3Abi,
                 functionName: 'totalSupply',
-                args: [pool.address as `0x${string}`],
-            } as const,
-        ])
-        .flat();
-
-    const configContracts = pools
-        .map((pool) => [
+                args: [pool as `0x${string}`],
+            },
             {
                 address: vault as `0x${string}`,
                 abi: vaultV3Abi,
                 functionName: 'getPoolConfig',
-                args: [pool.address as `0x${string}`],
-            } as const,
-        ])
-        .flat();
-
-    const protocolSwapFeeContracts = pools
-        .map((pool) => [
+                args: [pool as `0x${string}`],
+            },
+            {
+                address: vault as `0x${string}`,
+                abi: vaultV3Abi,
+                functionName: 'getPoolTokenInfo',
+                args: [pool as `0x${string}`],
+            },
             {
                 address: vault as `0x${string}`,
                 abi: vaultV3Abi,
-                functionName: 'getProtocolSwapFeePercentage',
-            } as const,
+                functionName: 'getPoolTokenRates',
+                args: [pool as `0x${string}`],
+            },
         ])
         .flat();
 
-    // TODO combine into one call
-    const totalSupplyResult = await client.multicall({ contracts: totalSupplyContracts, blockNumber: blockNumber });
-    const configResult = await client.multicall({ contracts: configContracts, blockNumber: blockNumber });
-    const protocolSwapFeeResult = await client.multicall({
-        contracts: protocolSwapFeeContracts,
-        blockNumber: blockNumber,
-    });
+    const results = await client.multicall({ contracts, blockNumber: blockNumber });
 
     // Parse the results
-    const parsedResults: Record<string, PoolData> = {};
-    pools.forEach((result, i) => {
-        if (
-            totalSupplyResult[i].status === 'success' &&
-            totalSupplyResult[i].result !== undefined &&
-            configResult[i].status === 'success' &&
-            configResult[i].result !== undefined &&
-            protocolSwapFeeResult[i].status === 'success' &&
-            protocolSwapFeeResult[i].result !== undefined
-        ) {
-            // parse the result here using the abi
-            const poolData = {
-                totalSupply: totalSupplyResult[i].result!,
-                swapFee: configResult[i].result!.staticSwapFeePercentage,
-                protocolSwapFeePercentage: 0n, // TODO can this be added to config?
-                isPoolPaused: configResult[i].result!.isPoolPaused,
-                isPoolInRecoveryMode: configResult[i].result!.isPoolInRecoveryMode,
-            } as PoolData;
+    const parsedResults = pools.map((pool, i) => {
```

Comment:
this whole parsing is so ugly. We need a way to abstract that properly.

Reply:
i agree that something similiar to multicaller would be great. I added this here: web3/multicaller-viem.ts – but making it work with the types is another challenge and i don't think we need to fight for this one. There are just few places where we are fetching onchain data, and they are always opinionated, i mean we always know what we are fetching, so it's easy to match inputs. if anything i'd consider that for a separate task.